testgrid/cmd/entomologist/main.go

```diff
@@ -0,0 +1,212 @@
+	fs.StringVar(&ghCredInput, "github-token", "", "GitHub OAuth credentials (optional)")
+	fs.StringVar(&o.organization, "github-org", "", "GitHub organization")
+	fs.StringVar(&o.repository, "github-repo", "", "GitHub repository")
+	fs.StringVar(&o.gcsPath, "output", "", "GCS output (gs://bucket)")
+	fs.StringVar(&o.gcsCredentials, "gcs-cred", "", "/path/to/service/account/credentials (as .json)")
+	fs.BoolVar(&o.oneshot, "oneshot", false, "Write proto once and exit instead of monitoring GitHub for changes")
+	fs.Parse(args)
+
+	o.ghCredentials = []byte(ghCredInput)
+
+	return o.validate()
+}
+
+func (o *options) validate() error {
+	if o.organization == "" || o.repository == "" {
+		return errors.New(fmt.Sprintf("%s/%s is not a valid GitHub location", o.organization, o.repository))
+	}
+	if o.gcsPath == "" || !strings.HasPrefix(o.gcsPath, "gs://") {
+		return errors.New("invalid or missing --output")
+	}
+	if o.gcsCredentials == "" {
+		return errors.New("--gcs-cred required for write operations")
+	}
+	return nil
+}
+
+func parseOptions() options {
+	var o options
+
+	if err := o.parseArgs(flag.CommandLine, os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("Invalid Flags")
+	}
+
+	return o
+}
+
+func main() {
+	opt := parseOptions()
+
+	getToken := func() []byte {
+		return opt.ghCredentials
+	}
+
+	ghct := github.NewClient(getToken, "", opt.ghProxyEndpoint)
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	client, _ := io.NewOpener(ctx, opt.gcsCredentials)
+	if client == nil {
+		logrus.Fatalf("Empty Credentials (at %s) not allowed for write operation", opt.gcsCredentials)
+	}
+
+	gcsFile := fmt.Sprintf("/bugs-%s-%s", opt.organization, opt.repository)
+	writer, _ := client.Writer(ctx, opt.gcsPath+gcsFile)
+	defer writer.Close()
+
+	// kick off goroutines
+	poll := func() {
+		issueState, err := pinIssues(ghct, opt)
+		if err != nil {
+			logrus.Fatalf("Could not get issues: %e", err)
+		}
+
+		out, _ := proto.Marshal(issueState)
+		n, _ := writer.Write(out)
+
+		logrus.Infof("Sending %d characters to GCS", n)
+	}
+
+	if opt.oneshot {
+		poll()
+		return
+	}
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	wait.Until(poll, defaultPollTime, stopCh)
+
+	sigTerm := make(chan os.Signal, 1)
+	signal.Notify(sigTerm, syscall.SIGTERM)
+	signal.Notify(sigTerm, syscall.SIGINT)
+	<-sigTerm
+
+	logrus.Info("Entomologist is closing...")
+}
+
+var tre = regexp.MustCompile(`(?m)^target:(.+)$`)
+
+func getIssues(client githubClient, o options) ([]github.Issue, error) {
+	issuesAndPRs, err := client.ListIssues(o.organization, o.repository)
+
+	issues := issuesAndPRs[:0]
+	for _, issue := range issuesAndPRs {
+		if !issue.IsPullRequest() {
+			issues = append(issues, issue)
+		}
+	}
+
+	return issues, err
+}
+
+// Algorithm for detecting and formatting test targets in issues
+func pinIssues(client githubClient, o options) (*issue_state.IssueState, error) {
+	issues, err := getIssues(client, o)
```

Comment:
Somehow here find a way to remember when was the last run to pass a last execution date to limit the scope of issues returned, see comment in github client.go